### PR TITLE
Add gestures for selecting multiple lines and multiple words

### PR DESCRIPTION
**Pull request by scholtzan**
This should fix https://github.com/google/xi-editor/issues/444

### CODE REVIEW

File: rust/core-lib/src/editor.rs [hunk 1]
```diff
@@ -904,6 +886,8 @@ impl Editor {
         let offset = self.view.line_col_to_offset(&self.text, line as usize, col as usize);
         match ty {
             GestureType::ToggleSel => self.view.toggle_sel(&self.text, offset),
+            GestureType::MultiLineSelect => self.view.select_line(&self.text, offset, line as usize, true),
+            GestureType::MultiWordSelect => self.view.select_word(&self.text, offset, true)
```
Comment (rscircus) [inline rust/core-lib/src/editor.rs]: Would it be more consistent to name these `GestureType`s to:
```
SelectLine
SelectWord
```?

And here comes the newbie question:
Why is this necessary at all?

File: rust/core-lib/src/rpc.rs [hunk 2]
```diff
@@ -289,6 +289,8 @@ pub struct EditCommand<T> {
 #[serde(rename_all = "snake_case")]
 pub enum GestureType {
     ToggleSel,
+    MultiLineSelect,
+    MultiWordSelect,
```
Comment (rscircus) [inline rust/core-lib/src/rpc.rs]: See above.

File: rust/core-lib/src/view.rs [hunk 3]
```diff
@@ -242,6 +243,55 @@ impl View {
...
+        let mut selection = Selection::new();
+
+        if multi_select {
+            selection = self.selection.clone();
+        }
+
+        selection.add_region(region);
+        self.set_selection(text, selection);
+
+        self.start_drag(offset, start, end);
```
Comment (rscircus) [inline rust/core-lib/src/view.rs]: So basically, 260-269 is the code to create the multi-selection.

Where is the `multi_select` flag set?

What would be a better option?

1. Create a function for this (which is repeated in the other function) and and add it to the original `editor.rs`. Or
2. Leave it as it is.
3. Create a function for the repeated identical part?

I'm wondering here...

Comment (cmyr) [reply rust/core-lib/src/view.rs]: @rscircus your intuition is correct, the client sends a different message to core depending on whether or not ctrl/cmd is present during the click.

@scholtzan I do agree that there's possibly room for some code reuse, but I think this is fine as-is. Most of the ways this could be cleaner would require some api additions around `Selection` and `SelRegion`.

File: rust/core-lib/src/editor.rs (hunk 1 above)
Comment (scholtzan) [reply rust/core-lib/src/editor.rs]: Regarding the why: I added support for multi selection of lines and words in xi-mac (see this pull-request: https://github.com/google/xi-mac/pull/159). So every time the gestures are executed in xi-mac (either by pressing <kbd>CMD</kbd> and clicking twice or three times) then xi-mac sends the RPCs to the core. The RPC messages contain as gesture type `multi_line_select` or `multi_word_select`. For message handling xi-editor will eventually deserialize the messages from JSON and `multi_line_select` will become `MultiLineSelect` and `multi_word_select` will become `MultiWordSelect`. Depending on the gesture type, which is determined at this specific point in the code, the corresponding actions (`self.view.select_line` or `self.view.select_word`) are performed. 

Regarding the name, these events are triggered when users press <kbd>CMD</kbd> + double click or <kbd>CMD</kbd> + triple click (on OSX <kbd>CMD</kbd>, on other platforms maybe <kbd>CTRL</kbd>). So I thought when users perform these gestures their intent is specifically to select multiple words/lines. 
`self.view.select_word` and `self.view.select_line` are also used for selecting single lines/words in `do_click`. The last parameter of these functions indicates whether multiple or single lines/words can be selected.

File: rust/core-lib/src/view.rs [hunk 4]
```diff
@@ -242,6 +243,55 @@ impl View {
...
+            end: end,
+            horiz: None,
+            affinity: Affinity::default(),
+        };
+
+        let mut selection = Selection::new();
+
+        if multi_select {
+            selection = self.selection.clone();
+        }
```
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: @cmyr Yes, the second one looks much better. Thanks for pointing that out. I updated it in my pull request.

File: rust/core-lib/src/view.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (scholtzan) [reply rust/core-lib/src/view.rs]: @cmyr If the `start_drag` is not there then the behaviour is a little weird. For example, if I add a word to my selection by pressing <kbd>CMD</kbd> + double click then the entire word gets selected. But the moment I start dragging, the part of the word before the location where I clicked gets deselected again. So I'd say `start_drag` should stay here.

File: rust/core-lib/src/editor.rs (hunk 1 above)
Comment (rscircus) [reply rust/core-lib/src/editor.rs]: @scholtzan thanks for the detailed explanation. Now it makes perfectly sense. I'll add the corresponding gestures to xi-gtk. :)
